_topics/en/op_checksigfromstack.md

```diff
@@ -0,0 +1,179 @@
+  to use `OP_CSFS` to verify an arbitrary message can enable several new
+  features for Bitcoin users:
+
+  - **Paying for signatures:** if Alice controls a private key that can
+    sign a transaction paying Bob, Bob can use OP_CSFS to trustlessly
+    offer to pay Alice for the signature he needs.  <!-- No source for
+    this claim, but it seems obvious to me. -->
+
+      More recently, protocols involving paying for signatures typically
+      [assume the use of adaptor signatures][kohen selling sigs] that
+      are more private and which use less block space.
+
+  - **Delegation:** Alice might want to delegate the authority to spend
+    her coins to Bob without explicitly creating an onchain transaction
+    transferring the coins to a 1-of-2 multisig between her and Bob.  If
+    Alice designs her scripts with this sort of delegation in mind, she
+    can put Bob's pubkey in a message and use OP_CSFS to prove that
+    she's delegated spending authority to that key.  <!-- Source:
+    "CHECKSIGFROMSTACK says you are going to push stuff on the stack,
+    hash it, and then you can verify arbitrary pubkey privkey pairs. You
+    can do this for delegation."
+    https://diyhpl.us/wiki/transcripts/bitcoin-core-dev-tech/2019-06-06-noinput-etc/
+    -->
+
+      An alternative approach that's more private, more flexible, and
+      more block-space efficient is [graftroot][], although this
+      requires a soft fork that has so far only been lightly discussed.
+
+  - **Oracles:** an oracle may agree to sign a message indicating the
+    outcome of an event, e.g. the name of the national team that wins a
+    sporting event.  Two or more users can then deposit money into a
+    script using OP_CSFS that will pay a different person depending on
+    which team the oracle indicates was the winner.  <!-- Source: "[...]
+    any kind of outside oracle data. Say we're betting on price, and we
+    have Bitstamp's key hardcoded in the script."
+    https://diyhpl.us/wiki/transcripts/bitcoin-core-dev-tech/2019-06-06-noinput-etc/
+    -->
+
+      More recent focus on oracle-moderated contracts involves using
+      [Discreet Log Contracts][dlcs] (DLCs), which can be more private
+      and more block-space efficient.
+
+  - **Double-spent protection bonds:** a service may promise to never
+    try to double spend its UTXOs in order to encourage its payees to
+    accept its unconfirmed transactions as reliable payments.  To
+    demonstrate its good faith, the service can use OP_CSFS to offer
+    payment of a bond to any user that can prove the same key was used
+    to create two different signatures for transactions spending the
+    same UTXO.  <!-- Source: "These new opcodes have several use cases,
+    including double-spent protection bonds"
+    https://web.archive.org/web/20160828061959/http://elementsproject.org/elements/opcodes
+    -->
+
+      This use of `OP_CSFS` can be compared to [single-show
```

Comment:
Is 'single-show signature' standard terminology for this? A similar concept is used in https://adiabat.github.io/dlc.pdf and https://eprint.iacr.org/2017/394.pdf but I can't find a standard name for it. I don't know if _single-show_ captures the concept. After all, I can show the same valid signatures many times!

Reply:
After a quick bit of research, it looks to me like the Official Standard is for everyone to make up their own term!  :-)  I think I got "single show" from gmaxwell:

- "single show signatures tend to be useful for other things in any case." http://gnusha.org/bitcoin-wizards/html/2015-11-24.log.html

- "What you're talking about is called a "single show signature"."  https://download.wpsoftware.net/bitcoin/wizards/2015-03-20.html

- "it's an old idea called a single show signature, you can see conversations about them back in 2013 in here." https://freenode.irclog.whitequark.org/bitcoin-wizards/2019-02-10

- "you can further improve things by setting up the signatures as single show," https://download.wpsoftware.net/bitcoin/wizards/2014-04-11.html

- "jeremyrubin: there are a number of ideas like the use of single show signatures that make sense 'academically' but I think don't make pratical sense because they're so exceptionally brittle" https://freenode.irclog.whitequark.org/bitcoin-wizards/2019-05-08

- (I have more citations from Maxwell's emails to the Bitcoin-Dev list)

Adam Back uses the term in this email: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2019-January/016620.html (although he uses "one show" in the email linked in this topic, so he's not consistent).

That's two knowledgeable cryptographers using the term, which is more than I can find agreeing on any other synonym.  If we find a better or more widely used term, I'm happy to change it later.

Reply:
That's certainly enough knowledgeable cryptographers to convince me!